adafruit_requests.py

```diff
@@ -357,10 +360,74 @@ def __init__(
         self._session_id = session_id
         self._last_response = None
 
+    def _build_boundary_data(self, files: dict):  # pylint: disable=too-many-locals
+        boundary_string = self._build_boundary_string()
+        content_length = 0
+        boundary_objects = []
+
+        for field_name, field_values in files.items():
+            file_name = field_values[0]
+            file_handle = field_values[1]
+
+            boundary_data = f"--{boundary_string}\r\n"
+            boundary_data += f'Content-Disposition: form-data; name="{field_name}"'
+            if file_name is not None:
+                boundary_data += f'; filename="{file_name}"'
+            boundary_data += "\r\n"
+            if len(field_values) >= 3:
+                file_content_type = field_values[2]
+                boundary_data += f"Content-Type: {file_content_type}\r\n"
+            if len(field_values) >= 4:
+                file_headers = field_values[3]
+                for file_header_key, file_header_value in file_headers.items():
+                    boundary_data += f"{file_header_key}: {file_header_value}\r\n"
+            boundary_data += "\r\n"
```

Comment:
You might want to try to reduce the amount of string concatenation, because that generates garbage and potentially causes fragmentation. For instance, lines 372 and 373 could be one string. You could collect all the strings in a list and then `join` them at the end.

Reply:
Since the concatenated string goes into `boundary_objects`, would it be best just to put the smaller strings in there?

Reply:
I wasn't sure whether it mattered how boundary objects was structured. I thought it might be a mix of strings and bytes. But if it's a bunch of chunks that are just sent out, and the structure doesn't matter, then the "objects" could be fragments.

Reply:
Initially I attempted to use multi-line string with triple quotes, but I found that it was causing trouble with string to bytestring conversions and switched to the current structure with lots of concatenating strings.

Reply:
This was changed to use the list instead of string concatenation with the latest commits merged in from Justin's PR to my branch.